### ISSUE TITLE
Bond-related error saving AMBER `prmtop` from ParmEd structure

### ISSUE THREAD

**Issue by slochower** [ParmEd/ParmEd#930]
I'm trying to use ParmEd to write an Amber `prmtop` file after establishing an OpenMM system and am encountering the error `AttributeError: 'NoneType' object has no attribute 'used'` (the full error can be seen at the notebook link below). I'm jumping through a few hoops with the system setup to convert from GAFF to smirnoff99Frosst. The path I'm taking is circuitous (some background is [here](https://github.com/openforcefield/openforcefield/issues/66)), but I believe it should work, and the steps can be found in this [notebook](https://github.com/slochower/openforcefield/blob/ad3cf9d26e8fc9de4af2726afe5fc2511b111c20/examples/host_guest_simulation/cyclodextrin-but.ipynb) with the source and generated files [here](https://github.com/slochower/openforcefield/tree/ad3cf9d26e8fc9de4af2726afe5fc2511b111c20/examples/host_guest_simulation/bcd-ben). 

Briefly, I use the OpenEye toolkit to read in `mol2` files for the host and guest, write an `xyz` file for the host (I do this to avoid an error with cyclic molecules that is likely the same as https://github.com/MobleyLab/benchmarksets/issues/60 and https://github.com/ParmEd/ParmEd/issues/898), then create a list of `OEMols` using the guest `mol2` file, the host `xyz` file, and SMILES strings for ions and water. I then create an OpenMM `system` from the `OEMols` and use a PDB of the solvated structure (with appropriate `CONECT` records) to set the topology and positions. I am able to successfully use the `system` to establish a ParmEd structure, but `structure.save` is where I encounter the error.

Based on some poking around, it seems (to me) that the structure does have the right atoms and bonds and bond types. What other steps should I take to try to tease out why the line `bond.type.used = True` is erroring?

- ParmEd 2.7.3
- OpenMM 7.1.1
- OpenEye 2017.Jun.1
- CentOS 7.4 with kernel 3.10.0-693.5.2.el7.x86_64

**Comment by davidlmobley** [comment]
@swails - can you give us some troubleshooting pointers? David is running into a bunch of issues with cyclic hosts that really are making it hard to use ParmEd, which is critical for getting our parameters into AMBER. 

@slochower - what would be the exact shortest way to reproduce this problem? I'm guessing if you provided a more minimal "just load this set of files and you'll see the problem" type example it would be easier for us to get help.

**Comment by swails** [comment]
Sorry for letting this slip through -- the spare time I have to devote to my "old" life has shrunk significantly in the past couple months.

The `AttributeError` you are getting will happen when there is a `bond.type` that is `None` (its default assignment unless it is explicitly assigned a type).  What this means is that there is at least one bond in the `Structure` that does not have an assigned type.

Looking at your notebook, this seems to be the case for SETTLEd waters.  The default for `rigidWater` in `ForceField.createSystem` is `True`, meaning that these DOF will be constrained, and OpenMM need not actually assign any parameters to these terms.

However, this doesn't satisfy Amber, which needs some dummy values.  Try adding the `rigidWater=False` value to the `ff.createSystem` call in that notebook and try again.  Hopefully then the `System` will contain at least *some* definition for the O-H and H-H water bonds that ParmEd can dump into the Amber prmtop file.

**Comment by slochower** [comment]
Thanks for the response and your thoughts, @swails, the delay is certainly understandable. I'll give this a shot soon and let you know if it works.

**Comment by davidlmobley** [comment]
@slochower - the `createSystem` for SMIROFF is not the same as for OpenMM, so we should check tomorrow or Saturday if we have this option in place and discuss with John. There should be an easy way to introduce bonds here if that's the issue.

**Comment by slochower** [comment]
I did not have time to check things thoroughly, but I wanted to drop a note that adding `rigidWater=False` to `createSystem` does not fix the issue. Maybe this is because of your comment @davidlmobley.

**Comment by swails** [comment]
Another thing you can do is create a dummy BondType and assign it to the missing bond types.  So something like this:

```python
bond_type = pmd.BondType(1.0, 1.0, list=struct.bond_types)
struct.bond_types.append(bond_type)
for bond in struct.bonds:
    if bond.type is None:
        bond.type = bond_type
```

Obviously you'll need to continue constraining these bonds in a simulation since they'll be meaningless in real life.  But it might help get you going.

**Comment by slochower** [comment]
@swails, I have confirmed that I can write a topology when reassigning the bonds, as you suggested. However, the `prmtop` file is a whopping 1.1 GB. I *think* this is a problem with ParmEd, because if I run a short test simulation using the `System` in OpenMM, it seems to run without an issue. The `prmtop` also causes segmentation faults with `sander` and `pmemd`, either because of the size or because of an overflow in the fixed width (I think) fields (see below).

I've been able to capture this is in a more simple example [notebook](https://github.com/slochower/smirnoff-host-guest/blob/3a319972d5b6fded1784eaad75d6bf5e754343a3/mol2-to-prmtop.ipynb). Briefly, cell 38 sets up the `System` and cell 45 loads it into a ParmEd structure. The resulting `prmtop` is too big upload to GitHub, but I've spit out the beginning and end of each `FLAG` section in the bottom of the notebook. Notice that there are huge numbers of interactions in `%FLAG NONBONDED_PARM_INDEX`. @davidlmobley should *all* atoms in the system, even those which are chemically identical, such as the water oxygens, be given their own atom type? I think that is what is happening here.

I just checked that the same behavior -- ParmEd writing a really large `prmtop` -- happens with a box of cyclohexane-ethanol, so it is not strictly a host-guest issue. [This](https://github.com/slochower/smirnoff-host-guest/blob/4d9f1946b6c8cfd1e2c82dc587ceef1ac5007e67/smiles-to-prmtop.ipynb) is probably as close to a minimal example that I can make for this situation (apologies that it depends on OpenEye tools). I didn't put it in the notebook, but the box has 239 residues (~3000 atoms) and the `prmtop` is 200+ MB.

Any thoughts on what is going on?

**Comment by davidlmobley** [comment]
I know why this happens with cyclohexane-ethanol; because of internal symmetries in cyclohexane, the atom mappings end up subtly different from one molecule to the next so you basically end up with every molecule being treated like a distinct entity. But I'm not sure why it would happen for a host-guest system. Or for water oxygens. I've run simulations in water before with SMIRNOFF and not generated huge files (though this makes me wonder if I wrote them to AMBER format...). 

Our new Topology format should fix some of these issues when it is ready, I think, though there should be something which can be done in the nearer term.

cc @jchodera

**Comment by slochower** [comment]
Thanks @davidlmobley. I was looking for a simple test case to reproduce it and just picked the cyclohexane-ethanol mixture without too much thought. Maybe I should have tried something simpler and seen if a box of pure methane, for example, exhibits the same behavior.

**Comment by davidlmobley** [comment]
Yes. Even more informative might be something with no internal symmetries at all, such as, say, 2-chlorophenol or something similar. 

To be more specific, there's an issue in AMBER I think where if multiple instances of a "molecule" do not have the same order of atoms, they get treated differently in a way which consumes a ton of space. And the way we assign parameters is we take the provided OpenEye molecules and match them onto the molecules in the provided topology file (matching the graph). In any case where multiple matches are equivalent, we just take one (perhaps the first one or some such). This means that for molecules with internal symmetries (which would even include methane) multiple matchings are possible and you won't always get the same order of atoms. This doesn't cause any problems in OpenMM, and problems are fairly minor in GROMACS, but it seems like in AMBER it's resulting in larger files than one might like.

**Comment by slochower** [comment]
I think I see what you're saying, but I don't believe atom ordering should be a problem for water, in the host-guest example.

**Comment by tuckerburgin** [comment]
Sorry if this answer is available elsewhere -- is there a solution for the exploding Amber topology file described in this thread yet? I would love to be able to build systems entirely within OpenMM and then export them to Amber for simulations/analysis, but haven't managed to successfully export to Amber-format topology without writing all those dummy bonds and getting an unreasonably huge topology file.
